# Ejercicio 103

Generar una representación de los elementos de un `DataFrame` con `applymap`.

## Solución:

In [1]:
import pandas as pd 
import numpy as np

In [142]:
numeros = [[3.1415, 2.717], [1.41, 1]]

df = pd.DataFrame(numeros)
df

0      1
0  3.1415  2.717
1  1.4100  1.000

In [143]:
df.applymap(lambda z:z**2)

0         1
0  9.869022  7.382089
1  1.988100  1.000000

# Ejercicio 104

Usar `groupby` para agrupar datos por una columna específica y contar la cantidad de elementos.

## Solución:

In [2]:
df = pd.read_csv('./employees.csv')
df

employ_id first_name last_name     email  phone_number   hire_date  \
0          100     Steven      King     SKING  515.123.4567  1987-06-17   
1          101      Neena   Kochhar  NKOCHHAR  515.123.4568  1987-06-18   
2          102        Lex   De Haan   LDEHAAN  515.123.4569  1987-06-19   
3          103  Alexander    Hunold   AHUNOLD  590.423.4567  1987-06-20   
4          104      Bruce     Ernst    BERNST  590.423.4568  1987-06-21   
..         ...        ...       ...       ...           ...         ...   
102        202        Pat       Fay      PFAY  603.123.6666  1987-09-27   
103        203      Susan    Mavris   SMAVRIS  515.123.7777  1987-09-28   
104        204    Hermann      Baer     HBAER  515.123.8888  1987-09-29   
105        205    Shelley   Higgins  SHIGGINS  515.123.8080  1987-09-30   
106        206    William     Gietz    WGIETZ  515.123.8181  1987-10-01   

         job_id  salary  commission_pct  manager_id  department_id  
0       AD_PRES   24000             NaN         NaN           90.0  
1         AD_VP   17000             NaN       100.0           90.0  
2         AD_VP   17000             NaN       100.0           90.0  
3       IT_PROG    9000             NaN       102.0           60.0  
4       IT_PROG    6000             NaN       103.0           60.0  
..          ...     ...             ...         ...            ...  
102      MK_REP    6000             NaN       201.0           20.0  
103      HR_REP    6500             NaN       101.0           40.0  
104      PR_REP   10000             NaN       101.0           70.0  
105      AC_MGR   12000             NaN       101.0          110.0  
106  AC_ACCOUNT    8300             NaN       205.0          110.0  

[107 rows x 11 columns]

In [145]:

# El código agrupa el DataFrame `df` por la columna 'first_name', luego calcula el tamaño de cada grupo y finalmente ordena los grupos en orden descendente según su tamaño.

df.groupby('first_name').size().sort_values(ascending=False)

first_name
Peter       3
David       3
John        3
Randall     2
Julia       2
           ..
Harrison    1
Guy         1
Girard      1
Gerald      1
Winston     1
Length: 91, dtype: int64

# Ejercicio 105

Usar la función `agg` para realizar varias operaciones por un eje de un `DataFrame`.

## Solución:

In [146]:
# `df['salary'].agg(np.mean)` está calculando la media (promedio) de los valores en la columna 'salary' del DataFrame `df`.
df['salary'].agg(np.mean)


6461.682242990654

In [147]:
# La función `df.agg(['max','min'])` se usa para calcular los valores máximo y mínimo para cada columna en un DataFrame `df`. Devuelve un nuevo DataFrame con los valores máximo y mínimo para cada columna.
df.agg(['max','min'])

employ_id first_name last_name    email        phone_number   hire_date  \
max        206    Winston   Zlotkey  WTAYLOR        650.509.4876  1987-10-01   
min        100       Adam      Abel   ABANDA  011.44.1343.329268  1987-06-17   

         job_id  salary  commission_pct  manager_id  department_id  
max      ST_MAN   24000             0.4       205.0          110.0  
min  AC_ACCOUNT    2100             0.1       100.0           10.0

In [148]:
# El código `df.agg({'salary':['min','mean'],'employ_id':'min'})` está realizando operaciones de agregación en el DataFrame `df`.
df.agg({'salary':['min','mean'],'employ_id':'min'})

salary  employ_id
min   2100.000000      100.0
mean  6461.682243        NaN

# Ejercicio 106

Eliminar registros que no cumplan con una condición sobre una columna específica.

## Solución:

In [149]:
df = df[df['salary'] <7000]
# El código `df = df[df['salary'] <7000]` está filtrando el DataFrame `df` para incluir solo filas donde la columna 'salary' tiene un valor inferior a 7000. El DataFrame resultante se vuelve a asignar al variable `df`.
df

In [150]:
df

employ_id first_name    last_name     email        phone_number  \
4          104      Bruce        Ernst    BERNST        590.423.4568   
5          105      David       Austin   DAUSTIN        590.423.4569   
6          106      Valli    Pataballa  VPATABAL        590.423.4560   
7          107      Diana      Lorentz  DLORENTZ        590.423.5567   
13         113       Luis         Popp     LPOPP        515.124.4567   
15         115  Alexander         Khoo     AKHOO        515.127.4562   
16         116     Shelli        Baida    SBAIDA        515.127.4563   
17         117      Sigal       Tobias   STOBIAS        515.127.4564   
18         118        Guy       Himuro   GHIMURO        515.127.4565   
19         119      Karen   Colmenares  KCOLMENA        515.127.4566   
23         123     Shanta      Vollman  SVOLLMAN        650.123.4234   
24         124      Kevin      Mourgos  KMOURGOS        650.123.5234   
25         125      Julia        Nayer    JNAYER        650.124.1214   
26         126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
27         127      James       Landry   JLANDRY        650.124.1334   
28         128     Steven       Markle   SMARKLE        650.124.1434   
29         129      Laura       Bissot   LBISSOT        650.124.5234   
30         130      Mozhe     Atkinson  MATKINSO        650.124.6234   
31         131      James       Marlow   JAMRLOW        650.124.7234   
32         132         TJ        Olson   TJOLSON        650.124.8234   
33         133      Jason       Mallin   JMALLIN        650.127.1934   
34         134    Michael       Rogers   MROGERS        650.127.1834   
35         135         Ki          Gee      KGEE        650.127.1734   
36         136      Hazel   Philtanker  HPHILTAN        650.127.1634   
37         137     Renske       Ladwig   RLADWIG        650.121.1234   
38         138    Stephen       Stiles   SSTILES        650.121.2034   
39         139       John          Seo      JSEO        650.121.2019   
40         140     Joshua        Patel    JPATEL        650.121.1834   
41         141     Trenna         Rajs     TRAJS        650.121.8009   
42         142     Curtis       Davies   CDAVIES        650.121.2994   
43         143    Randall        Matos    RMATOS        650.121.2874   
44         144      Peter       Vargas   PVARGAS        650.121.2004   
65         165      David          Lee      DLEE  011.44.1346.529268   
66         166     Sundar         Ande     SANDE  011.44.1346.629268   
67         167       Amit        Banda    ABANDA  011.44.1346.729268   
73         173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
79         179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
80         180    Winston       Taylor   WTAYLOR        650.507.9876   
81         181       Jean       Fleaur   JFLEAUR        650.507.9877   
82         182     Martha     Sullivan  MSULLIVA        650.507.9878   
83         183     Girard        Geoni    GGEONI        650.507.9879   
84         184    Nandita     Sarchand  NSARCHAN        650.509.1876   
85         185     Alexis         Bull     ABULL        650.509.2876   
86         186      Julia    Dellinger  JDELLING        650.509.3876   
87         187    Anthony       Cabrio   ACABRIO        650.509.4876   
88         188      Kelly        Chung    KCHUNG        650.505.1876   
89         189   Jennifer        Dilly    JDILLY        650.505.2876   
90         190    Timothy        Gates    TGATES        650.505.3876   
91         191    Randall      Perkins  RPERKINS        650.505.4876   
92         192      Sarah         Bell     SBELL        650.501.1876   
93         193    Britney      Everett  BEVERETT        650.501.2876   
94         194     Samuel       McCain   SMCCAIN        650.501.3876   
95         195      Vance        Jones    VJONES        650.501.4876   
96         196      Alana        Walsh    AWALSH        650.507.9811   
97         197      Kevin       Feeney   KFEE

# Ejercicio 107

Transformar los valores de un `DataFrame` con la función `transform`.

## Solución:

In [151]:
df.transform(func=lambda x:x*2)

employ_id          first_name               last_name             email  \
4          208          BruceBruce              ErnstErnst      BERNSTBERNST   
5          210          DavidDavid            AustinAustin    DAUSTINDAUSTIN   
6          212          ValliValli      PataballaPataballa  VPATABALVPATABAL   
7          214          DianaDiana          LorentzLorentz  DLORENTZDLORENTZ   
13         226            LuisLuis                PoppPopp        LPOPPLPOPP   
15         230  AlexanderAlexander                KhooKhoo        AKHOOAKHOO   
16         232        ShelliShelli              BaidaBaida      SBAIDASBAIDA   
17         234          SigalSigal            TobiasTobias    STOBIASSTOBIAS   
18         236              GuyGuy            HimuroHimuro    GHIMUROGHIMURO   
19         238          KarenKaren    ColmenaresColmenares  KCOLMENAKCOLMENA   
23         246        ShantaShanta          VollmanVollman  SVOLLMANSVOLLMAN   
24         248          KevinKevin          MourgosMourgos  KMOURGOSKMOURGOS   
25         250          JuliaJulia              NayerNayer      JNAYERJNAYER   
26         252          IreneIrene  MikkilineniMikkilineni  IMIKKILIIMIKKILI   
27         254          JamesJames            LandryLandry    JLANDRYJLANDRY   
28         256        StevenSteven            MarkleMarkle    SMARKLESMARKLE   
29         258          LauraLaura            BissotBissot    LBISSOTLBISSOT   
30         260          MozheMozhe        AtkinsonAtkinson  MATKINSOMATKINSO   
31         262          JamesJames            MarlowMarlow    JAMRLOWJAMRLOW   
32         264                TJTJ              OlsonOlson    TJOLSONTJOLSON   
33         266          JasonJason            MallinMallin    JMALLINJMALLIN   
34         268      MichaelMichael            RogersRogers    MROGERSMROGERS   
35         270                KiKi                  GeeGee          KGEEKGEE   
36         272          HazelHazel    PhiltankerPhiltanker  HPHILTANHPHILTAN   
37         274        RenskeRenske            LadwigLadwig    RLADWIGRLADWIG   
38         276      StephenStephen            StilesStiles    SSTILESSSTILES   
39         278            JohnJohn                  SeoSeo          JSEOJSEO   
40         280        JoshuaJoshua              PatelPatel      JPATELJPATEL   
41         282        TrennaTrenna                RajsRajs        TRAJSTRAJS   
42         284        CurtisCurtis            DaviesDavies    CDAVIESCDAVIES   
43         286      RandallRandall              MatosMatos      RMATOSRMATOS   
44         288          PeterPeter            VargasVargas    PVARGASPVARGAS   
65         330          DavidDavid                  LeeLee          DLEEDLEE   
66         332        SundarSundar                AndeAnde        SANDESANDE   
67         334            AmitAmit              BandaBanda      ABANDAABANDA   
73         346      SunditaSundita              KumarKumar      SKUMARSKUMAR   
79         358      CharlesCharles          JohnsonJohnson  CJOHNSONCJOHNSON   
80         360      WinstonWinston            TaylorTaylor    WTAYLORWTAYLOR   
81         362            JeanJean            FleaurFleaur    JFLEAURJFLEAUR   
82         364        MarthaMartha        SullivanSullivan  MSULLIVAMSULLIVA   
83         366        GirardGirard              GeoniGeoni      GGEONIGGEONI   
84         368      NanditaNandita        SarchandSarchand  NSARCHANNSARCHAN   
85         370        AlexisAlexis                BullBull        ABULLABULL   
86         372          JuliaJulia      DellingerDellinger  JDELLINGJDELLING   
87         374      AnthonyAnthony            CabrioCabrio    ACABRIOACABRIO   
88         376          KellyKelly              ChungChung      KCHUNGKCHUNG   
89         378    JenniferJennifer              DillyDilly      JDILLYJDILLY   
90         380      TimothyTimothy              GatesGates      TGATESTGATES   
91         382      RandallRandall          PerkinsPerkins  RPERKINSRPERKINS   
92   

# Ejercicio 108

Agrupar un conjunto de datos y luego calcular su promedio con `groupby` y `mean`.

## Solución:

In [152]:
df.groupby('last_name').size()

last_name
Ande           1
Atkinson       1
Austin         1
Baida          1
Banda          1
Bell           1
Bissot         1
Bull           1
Cabrio         1
Chung          1
Colmenares     1
Davies         1
Dellinger      1
Dilly          1
Ernst          1
Everett        1
Fay            1
Feeney         1
Fleaur         1
Gates          1
Gee            1
Geoni          1
Grant          1
Himuro         1
Johnson        1
Jones          1
Khoo           1
Kumar          1
Ladwig         1
Landry         1
Lee            1
Lorentz        1
Mallin         1
Markle         1
Marlow         1
Matos          1
Mavris         1
McCain         1
Mikkilineni    1
Mourgos        1
Nayer          1
OConnell       1
Olson          1
Pataballa      1
Patel          1
Perkins        1
Philtanker     1
Popp           1
Rajs           1
Rogers         1
Sarchand       1
Seo            1
Stiles         1
Sullivan       1
Taylor         1
Tobias         1
Vargas         1
Vollman        1
Wals

In [153]:
df.groupby('last_name')['salary'].agg(np.mean)

last_name
Ande           6400.0
Atkinson       2800.0
Austin         4800.0
Baida          2900.0
Banda          6200.0
Bell           4000.0
Bissot         3300.0
Bull           4100.0
Cabrio         3000.0
Chung          3800.0
Colmenares     2500.0
Davies         3100.0
Dellinger      3400.0
Dilly          3600.0
Ernst          6000.0
Everett        3900.0
Fay            6000.0
Feeney         3000.0
Fleaur         3100.0
Gates          2900.0
Gee            2400.0
Geoni          2800.0
Grant          2600.0
Himuro         2600.0
Johnson        6200.0
Jones          2800.0
Khoo           3100.0
Kumar          6100.0
Ladwig         3600.0
Landry         2400.0
Lee            6800.0
Lorentz        4200.0
Mallin         3300.0
Markle         2200.0
Marlow         2500.0
Matos          2600.0
Mavris         6500.0
McCain         3200.0
Mikkilineni    2700.0
Mourgos        5800.0
Nayer          3200.0
OConnell       2600.0
Olson          2100.0
Pataballa      4800.0
Patel          2500.0


# Ejercicio 109

Usar las funciones `all` y `any` sobre un objeto `DataFrame`.

## Solución:

In [154]:
(df != 100 ).all() # Todos los valores de la columna deben cumplir

employ_id          True
first_name         True
last_name          True
email              True
phone_number       True
hire_date          True
job_id             True
salary             True
commission_pct     True
manager_id        False
department_id     False
dtype: bool

In [155]:
(df == 6400).any() # Si al menos un registro de alguna columna cumple la condicion devuelve true en la columna

employ_id         False
first_name        False
last_name         False
email             False
phone_number      False
hire_date         False
job_id            False
salary             True
commission_pct    False
manager_id        False
department_id     False
dtype: bool

# Ejercicio 111

Obtener estadísticas básicas en un objeto `DataFrame` con la función `describe`.

## Solución:

In [156]:
df.describe()

employ_id       salary  commission_pct  manager_id  department_id
count   60.000000    60.000000             5.0   60.000000      60.000000
mean   155.033333  3691.666667             0.1  121.850000      51.000000
std     32.450205  1394.576865             0.0   15.075882      14.225782
min    104.000000  2100.000000             0.1  100.000000      10.000000
25%    127.750000  2675.000000             0.1  120.000000      50.000000
50%    142.500000  3150.000000             0.1  121.500000      50.000000
75%    187.250000  4200.000000             0.1  123.000000      50.000000
max    203.000000  6900.000000             0.1  201.000000     100.000000

# Ejercicio 114

Calcular el elemento mayor por cada fila y columna con la función `max`.

## Solución:

In [157]:
#df.max(axis) # Solo funciona para df numericos IDEM mean,median etc

In [158]:
df.query('salary == 6500 and employ_id == 203')

employ_id first_name last_name    email  phone_number   hire_date  \
103        203      Susan    Mavris  SMAVRIS  515.123.7777  1987-09-28   

     job_id  salary  commission_pct  manager_id  department_id  
103  HR_REP    6500             NaN       101.0           40.0

# Ejercicio 115

Reemplazar los NaN (nulo) con un valor específico usando la función `fillna`.

## Solución:

In [159]:
df.fillna(6,inplace=True)
df['commission_pct']

C:\Users\bruno\AppData\Local\Temp\ipykernel_9480\1485705238.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.fillna(6,inplace=True)


4      6.0
5      6.0
6      6.0
7      6.0
13     6.0
15     6.0
16     6.0
17     6.0
18     6.0
19     6.0
23     6.0
24     6.0
25     6.0
26     6.0
27     6.0
28     6.0
29     6.0
30     6.0
31     6.0
32     6.0
33     6.0
34     6.0
35     6.0
36     6.0
37     6.0
38     6.0
39     6.0
40     6.0
41     6.0
42     6.0
43     6.0
44     6.0
65     0.1
66     0.1
67     0.1
73     0.1
79     0.1
80     6.0
81     6.0
82     6.0
83     6.0
84     6.0
85     6.0
86     6.0
87     6.0
88     6.0
89     6.0
90     6.0
91     6.0
92     6.0
93     6.0
94     6.0
95     6.0
96     6.0
97     6.0
98     6.0
99     6.0
100    6.0
102    6.0
103    6.0
Name: commission_pct, dtype: float64

# Ejercicio 118

Convertir los índices de un `DataFrame` en una columna con la función `reset_index`.

## Solución:

In [160]:
df.reset_index()

index  employ_id first_name    last_name     email        phone_number  \
0       4        104      Bruce        Ernst    BERNST        590.423.4568   
1       5        105      David       Austin   DAUSTIN        590.423.4569   
2       6        106      Valli    Pataballa  VPATABAL        590.423.4560   
3       7        107      Diana      Lorentz  DLORENTZ        590.423.5567   
4      13        113       Luis         Popp     LPOPP        515.124.4567   
5      15        115  Alexander         Khoo     AKHOO        515.127.4562   
6      16        116     Shelli        Baida    SBAIDA        515.127.4563   
7      17        117      Sigal       Tobias   STOBIAS        515.127.4564   
8      18        118        Guy       Himuro   GHIMURO        515.127.4565   
9      19        119      Karen   Colmenares  KCOLMENA        515.127.4566   
10     23        123     Shanta      Vollman  SVOLLMAN        650.123.4234   
11     24        124      Kevin      Mourgos  KMOURGOS        650.123.5234   
12     25        125      Julia        Nayer    JNAYER        650.124.1214   
13     26        126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
14     27        127      James       Landry   JLANDRY        650.124.1334   
15     28        128     Steven       Markle   SMARKLE        650.124.1434   
16     29        129      Laura       Bissot   LBISSOT        650.124.5234   
17     30        130      Mozhe     Atkinson  MATKINSO        650.124.6234   
18     31        131      James       Marlow   JAMRLOW        650.124.7234   
19     32        132         TJ        Olson   TJOLSON        650.124.8234   
20     33        133      Jason       Mallin   JMALLIN        650.127.1934   
21     34        134    Michael       Rogers   MROGERS        650.127.1834   
22     35        135         Ki          Gee      KGEE        650.127.1734   
23     36        136      Hazel   Philtanker  HPHILTAN        650.127.1634   
24     37        137     Renske       Ladwig   RLADWIG        650.121.1234   
25     38        138    Stephen       Stiles   SSTILES        650.121.2034   
26     39        139       John          Seo      JSEO        650.121.2019   
27     40        140     Joshua        Patel    JPATEL        650.121.1834   
28     41        141     Trenna         Rajs     TRAJS        650.121.8009   
29     42        142     Curtis       Davies   CDAVIES        650.121.2994   
30     43        143    Randall        Matos    RMATOS        650.121.2874   
31     44        144      Peter       Vargas   PVARGAS        650.121.2004   
32     65        165      David          Lee      DLEE  011.44.1346.529268   
33     66        166     Sundar         Ande     SANDE  011.44.1346.629268   
34     67        167       Amit        Banda    ABANDA  011.44.1346.729268   
35     73        173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
36     79        179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
37     80        180    Winston       Taylor   WTAYLOR        650.507.9876   
38     81        181       Jean       Fleaur   JFLEAUR        650.507.9877   
39     82        182     Martha     Sullivan  MSULLIVA        650.507.9878   
40     83        183     Girard        Geoni    GGEONI        650.507.9879   
41     84        184    Nandita     Sarchand  NSARCHAN        650.509.1876   
42     85        185     Alexis         Bull     ABULL        650.509.2876   
43     86        186      Julia    Dellinger  JDELLING        650.509.3876   
44     87        187    Anthony       Cabrio   ACABRIO        650.509.4876   
45     88        188      Kelly        Chung    KCHUNG        650.505.1876   
46     89        189   Jennifer        Dilly    JDILLY        650.505.2876   
47     90        190    Timothy        Gates    TGATES        650.505.3876   
48     91        191    Randall      Perkins  RPERKINS        650.505.4876   
49     92        192      Sarah         Bell     SBELL        650.501.1876   
50     93        193    Br

# Ejercicio 119

Seleccionar registros que cumplan con una hora específica con `at_time`.

## Solución:

# Ejercicio 120

Usar la función `set_value` para cambiar un elemento específico en un `DataFrame`.

## Solución:

# Ejercicio 121

Calcular la productoria por cada fila y columna de un `DataFrame`.

$\prod_{i=1}^{n}k_{i}=k_1\times k_2 \times ...\times k_n$

## Solución:

In [161]:
#df.product(axis)  # Solo para numericos

# Ejercicio 122

Contar los elementos que son nulos (NaN) en un objeto `DataFrame`.

## Solución:

In [162]:
df.isnull().values.sum()

0

# Ejercicio 123

Redondear valores reales con la función `DataFrame.round`.

## Solución:

In [163]:
df.round()

employ_id first_name    last_name     email        phone_number  \
4          104      Bruce        Ernst    BERNST        590.423.4568   
5          105      David       Austin   DAUSTIN        590.423.4569   
6          106      Valli    Pataballa  VPATABAL        590.423.4560   
7          107      Diana      Lorentz  DLORENTZ        590.423.5567   
13         113       Luis         Popp     LPOPP        515.124.4567   
15         115  Alexander         Khoo     AKHOO        515.127.4562   
16         116     Shelli        Baida    SBAIDA        515.127.4563   
17         117      Sigal       Tobias   STOBIAS        515.127.4564   
18         118        Guy       Himuro   GHIMURO        515.127.4565   
19         119      Karen   Colmenares  KCOLMENA        515.127.4566   
23         123     Shanta      Vollman  SVOLLMAN        650.123.4234   
24         124      Kevin      Mourgos  KMOURGOS        650.123.5234   
25         125      Julia        Nayer    JNAYER        650.124.1214   
26         126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
27         127      James       Landry   JLANDRY        650.124.1334   
28         128     Steven       Markle   SMARKLE        650.124.1434   
29         129      Laura       Bissot   LBISSOT        650.124.5234   
30         130      Mozhe     Atkinson  MATKINSO        650.124.6234   
31         131      James       Marlow   JAMRLOW        650.124.7234   
32         132         TJ        Olson   TJOLSON        650.124.8234   
33         133      Jason       Mallin   JMALLIN        650.127.1934   
34         134    Michael       Rogers   MROGERS        650.127.1834   
35         135         Ki          Gee      KGEE        650.127.1734   
36         136      Hazel   Philtanker  HPHILTAN        650.127.1634   
37         137     Renske       Ladwig   RLADWIG        650.121.1234   
38         138    Stephen       Stiles   SSTILES        650.121.2034   
39         139       John          Seo      JSEO        650.121.2019   
40         140     Joshua        Patel    JPATEL        650.121.1834   
41         141     Trenna         Rajs     TRAJS        650.121.8009   
42         142     Curtis       Davies   CDAVIES        650.121.2994   
43         143    Randall        Matos    RMATOS        650.121.2874   
44         144      Peter       Vargas   PVARGAS        650.121.2004   
65         165      David          Lee      DLEE  011.44.1346.529268   
66         166     Sundar         Ande     SANDE  011.44.1346.629268   
67         167       Amit        Banda    ABANDA  011.44.1346.729268   
73         173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
79         179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
80         180    Winston       Taylor   WTAYLOR        650.507.9876   
81         181       Jean       Fleaur   JFLEAUR        650.507.9877   
82         182     Martha     Sullivan  MSULLIVA        650.507.9878   
83         183     Girard        Geoni    GGEONI        650.507.9879   
84         184    Nandita     Sarchand  NSARCHAN        650.509.1876   
85         185     Alexis         Bull     ABULL        650.509.2876   
86         186      Julia    Dellinger  JDELLING        650.509.3876   
87         187    Anthony       Cabrio   ACABRIO        650.509.4876   
88         188      Kelly        Chung    KCHUNG        650.505.1876   
89         189   Jennifer        Dilly    JDILLY        650.505.2876   
90         190    Timothy        Gates    TGATES        650.505.3876   
91         191    Randall      Perkins  RPERKINS        650.505.4876   
92         192      Sarah         Bell     SBELL        650.501.1876   
93         193    Britney      Everett  BEVERETT        650.501.2876   
94         194     Samuel       McCain   SMCCAIN        650.501.3876   
95         195      Vance        Jones    VJONES        650.501.4876   
96         196      Alana        Walsh    AWALSH        650.507.9811   
97         197      Kevin       Feeney   KFEE

# Ejercicio 125

Encontrar la cantidad de elementos únicos por cada fila y columna con `nunique`.

## Solución:

In [164]:
df.nunique() # PUEDO USAR EL PARAMETRO AXIS PARA VER SI EN UNA MISMA FILA LOS ORES SON UNICOS 

employ_id         60
first_name        54
last_name         60
email             60
phone_number      60
hire_date         60
job_id            10
salary            30
commission_pct     2
manager_id        14
department_id      8
dtype: int64

# Ejercicio 126

Reiniciar un índice numérico después de eliminar algunos registros con `reset_index`.

## Solución:

In [165]:
df.reset_index(inplace=True)


In [166]:
df.drop(columns='index',inplace=True)

C:\Users\bruno\AppData\Local\Temp\ipykernel_9480\2364643253.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns='index',inplace=True)


In [167]:
df.drop(columns='level_0',inplace=True)

KeyError: "['level_0'] not found in axis"

In [168]:
df

employ_id first_name    last_name     email        phone_number  \
0         104      Bruce        Ernst    BERNST        590.423.4568   
1         105      David       Austin   DAUSTIN        590.423.4569   
2         106      Valli    Pataballa  VPATABAL        590.423.4560   
3         107      Diana      Lorentz  DLORENTZ        590.423.5567   
4         113       Luis         Popp     LPOPP        515.124.4567   
5         115  Alexander         Khoo     AKHOO        515.127.4562   
6         116     Shelli        Baida    SBAIDA        515.127.4563   
7         117      Sigal       Tobias   STOBIAS        515.127.4564   
8         118        Guy       Himuro   GHIMURO        515.127.4565   
9         119      Karen   Colmenares  KCOLMENA        515.127.4566   
10        123     Shanta      Vollman  SVOLLMAN        650.123.4234   
11        124      Kevin      Mourgos  KMOURGOS        650.123.5234   
12        125      Julia        Nayer    JNAYER        650.124.1214   
13        126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
14        127      James       Landry   JLANDRY        650.124.1334   
15        128     Steven       Markle   SMARKLE        650.124.1434   
16        129      Laura       Bissot   LBISSOT        650.124.5234   
17        130      Mozhe     Atkinson  MATKINSO        650.124.6234   
18        131      James       Marlow   JAMRLOW        650.124.7234   
19        132         TJ        Olson   TJOLSON        650.124.8234   
20        133      Jason       Mallin   JMALLIN        650.127.1934   
21        134    Michael       Rogers   MROGERS        650.127.1834   
22        135         Ki          Gee      KGEE        650.127.1734   
23        136      Hazel   Philtanker  HPHILTAN        650.127.1634   
24        137     Renske       Ladwig   RLADWIG        650.121.1234   
25        138    Stephen       Stiles   SSTILES        650.121.2034   
26        139       John          Seo      JSEO        650.121.2019   
27        140     Joshua        Patel    JPATEL        650.121.1834   
28        141     Trenna         Rajs     TRAJS        650.121.8009   
29        142     Curtis       Davies   CDAVIES        650.121.2994   
30        143    Randall        Matos    RMATOS        650.121.2874   
31        144      Peter       Vargas   PVARGAS        650.121.2004   
32        165      David          Lee      DLEE  011.44.1346.529268   
33        166     Sundar         Ande     SANDE  011.44.1346.629268   
34        167       Amit        Banda    ABANDA  011.44.1346.729268   
35        173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
36        179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
37        180    Winston       Taylor   WTAYLOR        650.507.9876   
38        181       Jean       Fleaur   JFLEAUR        650.507.9877   
39        182     Martha     Sullivan  MSULLIVA        650.507.9878   
40        183     Girard        Geoni    GGEONI        650.507.9879   
41        184    Nandita     Sarchand  NSARCHAN        650.509.1876   
42        185     Alexis         Bull     ABULL        650.509.2876   
43        186      Julia    Dellinger  JDELLING        650.509.3876   
44        187    Anthony       Cabrio   ACABRIO        650.509.4876   
45        188      Kelly        Chung    KCHUNG        650.505.1876   
46        189   Jennifer        Dilly    JDILLY        650.505.2876   
47        190    Timothy        Gates    TGATES        650.505.3876   
48        191    Randall      Perkins  RPERKINS        650.505.4876   
49        192      Sarah         Bell     SBELL        650.501.1876   
50        193    Britney      Everett  BEVERETT        650.501.2876   
51        194     Samuel       McCain   SMCCAIN        650.501.3876   
52        195      Vance        Jones    VJONES        650.501.4876   
53        196      Alana        Walsh    AWALSH        650.507.9811   
54        197      Kevin       Feeney   KFEENEY        650.507.9822   
55        198     Donald    

# Ejercicio 127

Agregar un prefijo y postfijo a las columnas de un objeto `DataFrame`.

## Solución:

In [169]:
df.add_prefix(axis=0,prefix='fila_')

employ_id first_name    last_name     email        phone_number  \
fila_0         104      Bruce        Ernst    BERNST        590.423.4568   
fila_1         105      David       Austin   DAUSTIN        590.423.4569   
fila_2         106      Valli    Pataballa  VPATABAL        590.423.4560   
fila_3         107      Diana      Lorentz  DLORENTZ        590.423.5567   
fila_4         113       Luis         Popp     LPOPP        515.124.4567   
fila_5         115  Alexander         Khoo     AKHOO        515.127.4562   
fila_6         116     Shelli        Baida    SBAIDA        515.127.4563   
fila_7         117      Sigal       Tobias   STOBIAS        515.127.4564   
fila_8         118        Guy       Himuro   GHIMURO        515.127.4565   
fila_9         119      Karen   Colmenares  KCOLMENA        515.127.4566   
fila_10        123     Shanta      Vollman  SVOLLMAN        650.123.4234   
fila_11        124      Kevin      Mourgos  KMOURGOS        650.123.5234   
fila_12        125      Julia        Nayer    JNAYER        650.124.1214   
fila_13        126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
fila_14        127      James       Landry   JLANDRY        650.124.1334   
fila_15        128     Steven       Markle   SMARKLE        650.124.1434   
fila_16        129      Laura       Bissot   LBISSOT        650.124.5234   
fila_17        130      Mozhe     Atkinson  MATKINSO        650.124.6234   
fila_18        131      James       Marlow   JAMRLOW        650.124.7234   
fila_19        132         TJ        Olson   TJOLSON        650.124.8234   
fila_20        133      Jason       Mallin   JMALLIN        650.127.1934   
fila_21        134    Michael       Rogers   MROGERS        650.127.1834   
fila_22        135         Ki          Gee      KGEE        650.127.1734   
fila_23        136      Hazel   Philtanker  HPHILTAN        650.127.1634   
fila_24        137     Renske       Ladwig   RLADWIG        650.121.1234   
fila_25        138    Stephen       Stiles   SSTILES        650.121.2034   
fila_26        139       John          Seo      JSEO        650.121.2019   
fila_27        140     Joshua        Patel    JPATEL        650.121.1834   
fila_28        141     Trenna         Rajs     TRAJS        650.121.8009   
fila_29        142     Curtis       Davies   CDAVIES        650.121.2994   
fila_30        143    Randall        Matos    RMATOS        650.121.2874   
fila_31        144      Peter       Vargas   PVARGAS        650.121.2004   
fila_32        165      David          Lee      DLEE  011.44.1346.529268   
fila_33        166     Sundar         Ande     SANDE  011.44.1346.629268   
fila_34        167       Amit        Banda    ABANDA  011.44.1346.729268   
fila_35        173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
fila_36        179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
fila_37        180    Winston       Taylor   WTAYLOR        650.507.9876   
fila_38        181       Jean       Fleaur   JFLEAUR        650.507.9877   
fila_39        182     Martha     Sullivan  MSULLIVA        650.507.9878   
fila_40        183     Girard        Geoni    GGEONI        650.507.9879   
fila_41        184    Nandita     Sarchand  NSARCHAN        650.509.1876   
fila_42        185     Alexis         Bull     ABULL        650.509.2876   
fila_43        186      Julia    Dellinger  JDELLING        650.509.3876   
fila_44        187    Anthony       Cabrio   ACABRIO        650.509.4876   
fila_45        188      Kelly        Chung    KCHUNG        650.505.1876   
fila_46        189   Jennifer        Dilly    JDILLY        650.505.2876   
fila_47        190    Timothy        Gates    TGATES        650.505.3876   
fila_48        191    Randall      Perkins  RPERKINS        650.505.4876   
fila_49        192      Sarah         Bell     SBELL        650.501.1876   
fila_50        193    Britney      Everett  BEVERETT        650.501.2876   
fila_51        194     Samuel       McCain   SMCCAIN     

# Ejercicio 130

Eliminar registros duplicados de un `DataFrame` a partir de una columna.

## Solución:

In [ ]:
df.drop_duplicates('first_name')

employ_id first_name    last_name     email        phone_number  \
4          104      Bruce        Ernst    BERNST        590.423.4568   
5          105      David       Austin   DAUSTIN        590.423.4569   
6          106      Valli    Pataballa  VPATABAL        590.423.4560   
7          107      Diana      Lorentz  DLORENTZ        590.423.5567   
13         113       Luis         Popp     LPOPP        515.124.4567   
15         115  Alexander         Khoo     AKHOO        515.127.4562   
16         116     Shelli        Baida    SBAIDA        515.127.4563   
17         117      Sigal       Tobias   STOBIAS        515.127.4564   
18         118        Guy       Himuro   GHIMURO        515.127.4565   
19         119      Karen   Colmenares  KCOLMENA        515.127.4566   
23         123     Shanta      Vollman  SVOLLMAN        650.123.4234   
24         124      Kevin      Mourgos  KMOURGOS        650.123.5234   
25         125      Julia        Nayer    JNAYER        650.124.1214   
26         126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
27         127      James       Landry   JLANDRY        650.124.1334   
28         128     Steven       Markle   SMARKLE        650.124.1434   
29         129      Laura       Bissot   LBISSOT        650.124.5234   
30         130      Mozhe     Atkinson  MATKINSO        650.124.6234   
32         132         TJ        Olson   TJOLSON        650.124.8234   
33         133      Jason       Mallin   JMALLIN        650.127.1934   
34         134    Michael       Rogers   MROGERS        650.127.1834   
35         135         Ki          Gee      KGEE        650.127.1734   
36         136      Hazel   Philtanker  HPHILTAN        650.127.1634   
37         137     Renske       Ladwig   RLADWIG        650.121.1234   
38         138    Stephen       Stiles   SSTILES        650.121.2034   
39         139       John          Seo      JSEO        650.121.2019   
40         140     Joshua        Patel    JPATEL        650.121.1834   
41         141     Trenna         Rajs     TRAJS        650.121.8009   
42         142     Curtis       Davies   CDAVIES        650.121.2994   
43         143    Randall        Matos    RMATOS        650.121.2874   
44         144      Peter       Vargas   PVARGAS        650.121.2004   
66         166     Sundar         Ande     SANDE  011.44.1346.629268   
67         167       Amit        Banda    ABANDA  011.44.1346.729268   
73         173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
79         179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
80         180    Winston       Taylor   WTAYLOR        650.507.9876   
81         181       Jean       Fleaur   JFLEAUR        650.507.9877   
82         182     Martha     Sullivan  MSULLIVA        650.507.9878   
83         183     Girard        Geoni    GGEONI        650.507.9879   
84         184    Nandita     Sarchand  NSARCHAN        650.509.1876   
85         185     Alexis         Bull     ABULL        650.509.2876   
87         187    Anthony       Cabrio   ACABRIO        650.509.4876   
88         188      Kelly        Chung    KCHUNG        650.505.1876   
89         189   Jennifer        Dilly    JDILLY        650.505.2876   
90         190    Timothy        Gates    TGATES        650.505.3876   
92         192      Sarah         Bell     SBELL        650.501.1876   
93         193    Britney      Everett  BEVERETT        650.501.2876   
94         194     Samuel       McCain   SMCCAIN        650.501.3876   
95         195      Vance        Jones    VJONES        650.501.4876   
96         196      Alana        Walsh    AWALSH        650.507.9811   
98         198     Donald     OConnell  DOCONNEL        650.507.9833   
99         199    Douglas        Grant    DGRANT        650.507.9844   
102        202        Pat          Fay      PFAY        603.123.6666   
103        203      Susan       Mavris   SMAVRIS        515.123.7777   

      hire_date      job_id  salary  commiss

In [ ]:
df['salary'].idxmax()

13

In [ ]:
df

employ_id first_name    last_name     email        phone_number  \
4          104      Bruce        Ernst    BERNST        590.423.4568   
5          105      David       Austin   DAUSTIN        590.423.4569   
6          106      Valli    Pataballa  VPATABAL        590.423.4560   
7          107      Diana      Lorentz  DLORENTZ        590.423.5567   
13         113       Luis         Popp     LPOPP        515.124.4567   
15         115  Alexander         Khoo     AKHOO        515.127.4562   
16         116     Shelli        Baida    SBAIDA        515.127.4563   
17         117      Sigal       Tobias   STOBIAS        515.127.4564   
18         118        Guy       Himuro   GHIMURO        515.127.4565   
19         119      Karen   Colmenares  KCOLMENA        515.127.4566   
23         123     Shanta      Vollman  SVOLLMAN        650.123.4234   
24         124      Kevin      Mourgos  KMOURGOS        650.123.5234   
25         125      Julia        Nayer    JNAYER        650.124.1214   
26         126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
27         127      James       Landry   JLANDRY        650.124.1334   
28         128     Steven       Markle   SMARKLE        650.124.1434   
29         129      Laura       Bissot   LBISSOT        650.124.5234   
30         130      Mozhe     Atkinson  MATKINSO        650.124.6234   
31         131      James       Marlow   JAMRLOW        650.124.7234   
32         132         TJ        Olson   TJOLSON        650.124.8234   
33         133      Jason       Mallin   JMALLIN        650.127.1934   
34         134    Michael       Rogers   MROGERS        650.127.1834   
35         135         Ki          Gee      KGEE        650.127.1734   
36         136      Hazel   Philtanker  HPHILTAN        650.127.1634   
37         137     Renske       Ladwig   RLADWIG        650.121.1234   
38         138    Stephen       Stiles   SSTILES        650.121.2034   
39         139       John          Seo      JSEO        650.121.2019   
40         140     Joshua        Patel    JPATEL        650.121.1834   
41         141     Trenna         Rajs     TRAJS        650.121.8009   
42         142     Curtis       Davies   CDAVIES        650.121.2994   
43         143    Randall        Matos    RMATOS        650.121.2874   
44         144      Peter       Vargas   PVARGAS        650.121.2004   
65         165      David          Lee      DLEE  011.44.1346.529268   
66         166     Sundar         Ande     SANDE  011.44.1346.629268   
67         167       Amit        Banda    ABANDA  011.44.1346.729268   
73         173    Sundita        Kumar    SKUMAR  011.44.1343.329268   
79         179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
80         180    Winston       Taylor   WTAYLOR        650.507.9876   
81         181       Jean       Fleaur   JFLEAUR        650.507.9877   
82         182     Martha     Sullivan  MSULLIVA        650.507.9878   
83         183     Girard        Geoni    GGEONI        650.507.9879   
84         184    Nandita     Sarchand  NSARCHAN        650.509.1876   
85         185     Alexis         Bull     ABULL        650.509.2876   
86         186      Julia    Dellinger  JDELLING        650.509.3876   
87         187    Anthony       Cabrio   ACABRIO        650.509.4876   
88         188      Kelly        Chung    KCHUNG        650.505.1876   
89         189   Jennifer        Dilly    JDILLY        650.505.2876   
90         190    Timothy        Gates    TGATES        650.505.3876   
91         191    Randall      Perkins  RPERKINS        650.505.4876   
92         192      Sarah         Bell     SBELL        650.501.1876   
93         193    Britney      Everett  BEVERETT        650.501.2876   
94         194     Samuel       McCain   SMCCAIN        650.501.3876   
95         195      Vance        Jones    VJONES        650.501.4876   
96         196      Alana        Walsh    AWALSH        650.507.9811   
97         197      Kevin       Feeney   KFEE

# Ejercicio 131

Cambiar el orden de las filas de forma aleatoria con la función `sample`.

## Solución:

In [ ]:
df.sample(n=df.shape[0])

employ_id first_name    last_name     email        phone_number  \
41         141     Trenna         Rajs     TRAJS        650.121.8009   
92         192      Sarah         Bell     SBELL        650.501.1876   
32         132         TJ        Olson   TJOLSON        650.124.8234   
79         179    Charles      Johnson  CJOHNSON  011.44.1644.429262   
40         140     Joshua        Patel    JPATEL        650.121.1834   
34         134    Michael       Rogers   MROGERS        650.127.1834   
36         136      Hazel   Philtanker  HPHILTAN        650.127.1634   
84         184    Nandita     Sarchand  NSARCHAN        650.509.1876   
35         135         Ki          Gee      KGEE        650.127.1734   
90         190    Timothy        Gates    TGATES        650.505.3876   
28         128     Steven       Markle   SMARKLE        650.124.1434   
88         188      Kelly        Chung    KCHUNG        650.505.1876   
94         194     Samuel       McCain   SMCCAIN        650.501.3876   
33         133      Jason       Mallin   JMALLIN        650.127.1934   
100        200   Jennifer       Whalen   JWHALEN        515.123.4444   
13         113       Luis         Popp     LPOPP        515.124.4567   
96         196      Alana        Walsh    AWALSH        650.507.9811   
29         129      Laura       Bissot   LBISSOT        650.124.5234   
30         130      Mozhe     Atkinson  MATKINSO        650.124.6234   
18         118        Guy       Himuro   GHIMURO        515.127.4565   
65         165      David          Lee      DLEE  011.44.1346.529268   
97         197      Kevin       Feeney   KFEENEY        650.507.9822   
99         199    Douglas        Grant    DGRANT        650.507.9844   
83         183     Girard        Geoni    GGEONI        650.507.9879   
6          106      Valli    Pataballa  VPATABAL        590.423.4560   
87         187    Anthony       Cabrio   ACABRIO        650.509.4876   
17         117      Sigal       Tobias   STOBIAS        515.127.4564   
103        203      Susan       Mavris   SMAVRIS        515.123.7777   
25         125      Julia        Nayer    JNAYER        650.124.1214   
15         115  Alexander         Khoo     AKHOO        515.127.4562   
27         127      James       Landry   JLANDRY        650.124.1334   
24         124      Kevin      Mourgos  KMOURGOS        650.123.5234   
91         191    Randall      Perkins  RPERKINS        650.505.4876   
44         144      Peter       Vargas   PVARGAS        650.121.2004   
23         123     Shanta      Vollman  SVOLLMAN        650.123.4234   
43         143    Randall        Matos    RMATOS        650.121.2874   
93         193    Britney      Everett  BEVERETT        650.501.2876   
85         185     Alexis         Bull     ABULL        650.509.2876   
39         139       John          Seo      JSEO        650.121.2019   
38         138    Stephen       Stiles   SSTILES        650.121.2034   
4          104      Bruce        Ernst    BERNST        590.423.4568   
26         126      Irene  Mikkilineni  IMIKKILI        650.124.1224   
86         186      Julia    Dellinger  JDELLING        650.509.3876   
42         142     Curtis       Davies   CDAVIES        650.121.2994   
37         137     Renske       Ladwig   RLADWIG        650.121.1234   
5          105      David       Austin   DAUSTIN        590.423.4569   
16         116     Shelli        Baida    SBAIDA        515.127.4563   
31         131      James       Marlow   JAMRLOW        650.124.7234   
82         182     Martha     Sullivan  MSULLIVA        650.507.9878   
7          107      Diana      Lorentz  DLORENTZ        590.423.5567   
67         167       Amit        Banda    ABANDA  011.44.1346.729268   
89         189   Jennifer        Dilly    JDILLY        650.505.2876   
98         198     Donald     OConnell  DOCONNEL        650.507.9833   
66         166     Sundar         Ande     SANDE  011.44.1346.629268   
19         119      Karen   Colmenares  KCOLM

# Ejercicio 132

Verificar con `equals` si dos objetos `DataFrame` contienen los mismos elementos.

## Solución:

In [ ]:
# df.eq(df2)

# Ejercicio 133

Convertir una columna de tipo cadena de caracteres a fecha con `to_datetime`.

## Solución:

# Ejercicio 134

Usar `first` y `last` para obtener los primeros y últimos registros para una cantidad de días arbitraria

## Solución:

# Ejercicio 135

Renombrar una columna de un `DataFrame` con la función `rename`.

## Solución:

In [ ]:
df.rename(columns={'hire_date':'fecha'},inplace=True)

C:\Users\bruno\AppData\Local\Temp\ipykernel_9480\2801746904.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns={'hire_date':'fecha'},inplace=True)


# Ejercicio 136

Obtener los datos de una columna como una lista Python con `tolist`.

## Solución:

In [ ]:
df['fecha'].to_list()

[<bound method Series.to_timestamp of 4      1987-06-21
 5      1987-06-22
 6      1987-06-23
 7      1987-06-24
 13     1987-06-30
 15     1987-07-02
 16     1987-07-03
 17     1987-07-04
 18     1987-07-05
 19     1987-07-06
 23     1987-07-10
 24     1987-07-11
 25     1987-07-12
 26     1987-07-13
 27     1987-07-14
 28     1987-07-15
 29     1987-07-16
 30     1987-07-17
 31     1987-07-18
 32     1987-07-19
 33     1987-07-20
 34     1987-07-21
 35     1987-07-22
 36     1987-07-23
 37     1987-07-24
 38     1987-07-25
 39     1987-07-26
 40     1987-07-27
 41     1987-07-28
 42     1987-07-29
 43     1987-07-30
 44     1987-07-31
 65     1987-08-21
 66     1987-08-22
 67     1987-08-23
 73     1987-08-29
 79     1987-09-04
 80     1987-09-05
 81     1987-09-06
 82     1987-09-07
 83     1987-09-08
 84     1987-09-09
 85     1987-09-10
 86     1987-09-11
 87     1987-09-12
 88     1987-09-13
 89     1987-09-14
 90     1987-09-15
 91     1987-09-16
 92     1987-09-17
 93     1987-

# Ejercicio 137

Seleccionar filas de un `DataFrame` a partir su posición con `take`.

## Solución:

In [ ]:
df.take(indices=[0,1]) # DEVUELVE LAS FILAS POR POSICION SEGUN EL ARRAY PASADO

employ_id first_name last_name    email  phone_number  \
4        104      Bruce     Ernst   BERNST  590.423.4568   
5        105      David    Austin  DAUSTIN  590.423.4569   

                                               fecha   job_id  salary  \
4  <bound method Series.to_timestamp of 4      19...  IT_PROG    6000   
5  <bound method Series.to_timestamp of 4      19...  IT_PROG    4800   

   commission_pct  manager_id  department_id  
4             NaN       103.0           60.0  
5             NaN       103.0           60.0

# Ejercicio 138

Seleccionar una columna de un `DataFrame` como índice con `set_index`.

## Solución:

In [ ]:
df.set_index('employ_id')

first_name    last_name     email        phone_number  \
employ_id                                                         
104            Bruce        Ernst    BERNST        590.423.4568   
105            David       Austin   DAUSTIN        590.423.4569   
106            Valli    Pataballa  VPATABAL        590.423.4560   
107            Diana      Lorentz  DLORENTZ        590.423.5567   
113             Luis         Popp     LPOPP        515.124.4567   
115        Alexander         Khoo     AKHOO        515.127.4562   
116           Shelli        Baida    SBAIDA        515.127.4563   
117            Sigal       Tobias   STOBIAS        515.127.4564   
118              Guy       Himuro   GHIMURO        515.127.4565   
119            Karen   Colmenares  KCOLMENA        515.127.4566   
123           Shanta      Vollman  SVOLLMAN        650.123.4234   
124            Kevin      Mourgos  KMOURGOS        650.123.5234   
125            Julia        Nayer    JNAYER        650.124.1214   
126            Irene  Mikkilineni  IMIKKILI        650.124.1224   
127            James       Landry   JLANDRY        650.124.1334   
128           Steven       Markle   SMARKLE        650.124.1434   
129            Laura       Bissot   LBISSOT        650.124.5234   
130            Mozhe     Atkinson  MATKINSO        650.124.6234   
131            James       Marlow   JAMRLOW        650.124.7234   
132               TJ        Olson   TJOLSON        650.124.8234   
133            Jason       Mallin   JMALLIN        650.127.1934   
134          Michael       Rogers   MROGERS        650.127.1834   
135               Ki          Gee      KGEE        650.127.1734   
136            Hazel   Philtanker  HPHILTAN        650.127.1634   
137           Renske       Ladwig   RLADWIG        650.121.1234   
138          Stephen       Stiles   SSTILES        650.121.2034   
139             John          Seo      JSEO        650.121.2019   
140           Joshua        Patel    JPATEL        650.121.1834   
141           Trenna         Rajs     TRAJS        650.121.8009   
142           Curtis       Davies   CDAVIES        650.121.2994   
143          Randall        Matos    RMATOS        650.121.2874   
144            Peter       Vargas   PVARGAS        650.121.2004   
165            David          Lee      DLEE  011.44.1346.529268   
166           Sundar         Ande     SANDE  011.44.1346.629268   
167             Amit        Banda    ABANDA  011.44.1346.729268   
173          Sundita        Kumar    SKUMAR  011.44.1343.329268   
179          Charles      Johnson  CJOHNSON  011.44.1644.429262   
180          Winston       Taylor   WTAYLOR        650.507.9876   
181             Jean       Fleaur   JFLEAUR        650.507.9877   
182           Martha     Sullivan  MSULLIVA        650.507.9878   
183           Girard        Geoni    GGEONI        650.507.9879   
184          Nandita     Sarchand  NSARCHAN        650.509.1876   
185           Alexis         Bull     ABULL        650.509.2876   
186            Julia    Dellinger  JDELLING        650.509.3876   
187          Anthony       Cabrio   ACABRIO        650.509.4876   
188            Kelly        Chung    KCHUNG        650.505.1876   
189         Jennifer        Dilly    JDILLY        650.505.2876   
190          Timothy        Gates    TGATES        650.505.3876   
191          Randall      Perkins  RPERKINS        650.505.4876   
192            Sarah         Bell     SBELL        650.501.1876   
193          Britney      Everett  BEVERETT        650.501.2876   
194           Samuel       McCain   SMCCAIN        650.501.3876   
195            Vance        Jones    VJONES        650.501.4876   
196            Alana        Walsh    AWALSH        650.507.9811   
197            Kevin       Feeney   KFEENEY        650.507.9822   
198           Donald     OConnell  DOCONNEL        650.507.9833   
199          Douglas        Grant    DGRANT        650.507.9844   
200         Jennifer       Whalen   JWHALEN        515.12

# Ejercicio 139

Truncar un `DataFrame` por filas y columnas con la función `truncate`.

## Solución:

In [ ]:
df.truncate(before=5,after=10) # TRUNCA LAS FILAS DESPUES  DEL INDICE AFTER
                               # Y ANTES DEL INDICE BEFORE

employ_id first_name  last_name     email  phone_number  \
5        105      David     Austin   DAUSTIN  590.423.4569   
6        106      Valli  Pataballa  VPATABAL  590.423.4560   
7        107      Diana    Lorentz  DLORENTZ  590.423.5567   

                                               fecha   job_id  salary  \
5  <bound method Series.to_timestamp of 4      19...  IT_PROG    4800   
6  <bound method Series.to_timestamp of 4      19...  IT_PROG    4800   
7  <bound method Series.to_timestamp of 4      19...  IT_PROG    4200   

   commission_pct  manager_id  department_id  
5             NaN       103.0           60.0  
6             NaN       103.0           60.0  
7             NaN       103.0           60.0

 # Ejercicio 140
 
 Crear un `DataFrame` a partir de arreglos `NumPy` y crear columnas e índices.
 
 ## Solución

In [3]:
array = np.random.randint(5,2000,(8,5))
array

array([[ 710, 1600,  980, 1347, 1537],
       [1581,  243,  114,  940, 1068],
       [1699, 1470,  196, 1023,  811],
       [ 857,  342, 1218,  650, 1610],
       [1881,  774,  347, 1232,  495],
       [1728, 1476,  129,  568,  226],
       [ 916, 1144,  452,  942,  414],
       [ 504,  371,  433, 1064, 1968]])

In [4]:
columnas=['x','y','z','u','v']
indices = ['a','b','c','d','e','e','f','g']

In [5]:
df_array = pd.DataFrame(data=array,columns=columnas,index=indices)
df_array

x     y     z     u     v
a   710  1600   980  1347  1537
b  1581   243   114   940  1068
c  1699  1470   196  1023   811
d   857   342  1218   650  1610
e  1881   774   347  1232   495
e  1728  1476   129   568   226
f   916  1144   452   942   414
g   504   371   433  1064  1968

In [173]:
print(f'Maximo por columna: \n{df_array.max()}')
print(f'Promedio por columna: \n{df_array.mean()}')
print(f'Valor medio por columna: \n{df_array.median()}')
print(f'Desviacion estandar por columna: \n{df_array.std()}')


Maximo por columna: 
x    1414
y    1879
z    1952
u    1999
v    1310
dtype: int32
Promedio por columna: 
x     584.500
y     957.625
z     998.875
u    1211.000
v     740.375
dtype: float64
Valor medio por columna: 
x     469.0
y     955.5
z    1125.0
u    1494.5
v     811.5
dtype: float64
Desviacion estandar por columna: 
x    487.873227
y    552.586112
z    749.524885
u    781.171465
v    400.305575
dtype: float64


# Ejercicio 141

Agregar una nueva columna a un `DataFrame` por medio de la función `assign`.

## Solución:

In [174]:
df_array.assign(w = lambda x:x['x'] **2) # Creo a columna w resoltado de la funcion labda que es la columna x al cuadrado

x     y     z     u     v        w
a   152  1101   176   506   376    23104
b  1175   810  1876    72   900  1380625
c    28   303   146   394   744      784
d   689  1879  1245  1789   276   474721
e   503   425  1952  1894  1310   253009
e   435   467  1390  1834  1168   189225
f   280  1294   201  1999   879    78400
g  1414  1382  1005  1200   270  1999396

# Ejercicio 142

Seleccionar el índice de una columna de un `DataFrame` con el elemento mayor.

## Solución:

In [193]:
df_array['y'].argmax()

3

In [178]:
df.columns

Index(['employ_id', 'first_name', 'last_name', 'email', 'phone_number',
       'hire_date', 'job_id', 'salary', 'commission_pct', 'manager_id',
       'department_id'],
      dtype='object')

In [182]:
result = df.loc[:, df.columns.str.en('me')]
result

first_name    last_name  department_id
0       Bruce        Ernst           60.0
1       David       Austin           60.0
2       Valli    Pataballa           60.0
3       Diana      Lorentz           60.0
4        Luis         Popp          100.0
5   Alexander         Khoo           30.0
6      Shelli        Baida           30.0
7       Sigal       Tobias           30.0
8         Guy       Himuro           30.0
9       Karen   Colmenares           30.0
10     Shanta      Vollman           50.0
11      Kevin      Mourgos           50.0
12      Julia        Nayer           50.0
13      Irene  Mikkilineni           50.0
14      James       Landry           50.0
15     Steven       Markle           50.0
16      Laura       Bissot           50.0
17      Mozhe     Atkinson           50.0
18      James       Marlow           50.0
19         TJ        Olson           50.0
20      Jason       Mallin           50.0
21    Michael       Rogers           50.0
22         Ki          Gee           50.0
23      Hazel   Philtanker           50.0
24     Renske       Ladwig           50.0
25    Stephen       Stiles           50.0
26       John          Seo           50.0
27     Joshua        Patel           50.0
28     Trenna         Rajs           50.0
29     Curtis       Davies           50.0
30    Randall        Matos           50.0
31      Peter       Vargas           50.0
32      David          Lee           80.0
33     Sundar         Ande           80.0
34       Amit        Banda           80.0
35    Sundita        Kumar           80.0
36    Charles      Johnson           80.0
37    Winston       Taylor           50.0
38       Jean       Fleaur           50.0
39     Martha     Sullivan           50.0
40     Girard        Geoni           50.0
41    Nandita     Sarchand           50.0
42     Alexis         Bull           50.0
43      Julia    Dellinger           50.0
44    Anthony       Cabrio           50.0
45      Kelly        Chung           50.0
46   Jennifer        Dilly           50.0
47    Timothy        Gates           50.0
48    Randall      Perkins           50.0
49      Sarah         Bell           50.0
50    Britney      Everett           50.0
51     Samuel       McCain           50.0
52      Vance        Jones           50.0
53      Alana        Walsh           50.0
54      Kevin       Feeney           50.0
55     Donald     OConnell           50.0
56    Douglas        Grant           50.0
57   Jennifer       Whalen           10.0
58        Pat          Fay           20.0
59      Susan       Mavris           40.0

In [194]:
df_array.sum(axis=1)

a    2311
b    4833
c    1615
d    5878
e    6084
e    5294
f    4653
g    5271
dtype: int64

# Ejercicio 143

Comprobar si una columna existe en un `DataFrame` con el operador Python `in`.

## Solución:

In [7]:
'employ_id' in df

True

# Ejercicio 145

Reemplazar valores en un `DataFrame` con la función `replace`.

## Solución:

In [9]:
df_array.replace(710,900)

x     y     z     u     v
a   900  1600   980  1347  1537
b  1581   243   114   940  1068
c  1699  1470   196  1023   811
d   857   342  1218   650  1610
e  1881   774   347  1232   495
e  1728  1476   129   568   226
f   916  1144   452   942   414
g   504   371   433  1064  1968

# Ejercicio 146

Obtener filas de un `DataFrame` con `iloc`.

## Solución:

In [10]:
df_array.iloc[[0,1,2]]

x     y    z     u     v
a   710  1600  980  1347  1537
b  1581   243  114   940  1068
c  1699  1470  196  1023   811

# Ejercicio 147

Obtener los tipos de datos de las columnas de un `DataFrame` con `dtypes`.

## Solución:

In [11]:
df.dtypes

employ_id           int64
first_name         object
last_name          object
email              object
phone_number       object
hire_date          object
job_id             object
salary              int64
commission_pct    float64
manager_id        float64
department_id     float64
dtype: object

# Ejercicio 148

Ordenar los valores de las columnas de un `DataFrame` con `sort_values`.

## Solución:

In [15]:
df_array.sort_values(by='x')

x     y     z     u     v
g   504   371   433  1064  1968
a   710  1600   980  1347  1537
d   857   342  1218   650  1610
f   916  1144   452   942   414
b  1581   243   114   940  1068
c  1699  1470   196  1023   811
e  1728  1476   129   568   226
e  1881   774   347  1232   495

# Ejercicio 149

Agregar datos a un `DataFrame` vacío con la función `append`.

## Solución:

In [17]:
df_nuevo = pd.DataFrame(index=[1,2,3,4,5,6],columns=['col1','col2','col3'])

# Ejercicio 150

Ordenar los índices de un `DataFrame` con la función `sort_index`.

## Solución:

In [24]:
df_array.sort_index()

x     y     z     u     v
a   710  1600   980  1347  1537
b  1581   243   114   940  1068
c  1699  1470   196  1023   811
d   857   342  1218   650  1610
e  1881   774   347  1232   495
e  1728  1476   129   568   226
f   916  1144   452   942   414
g   504   371   433  1064  1968

# Ejercicio 151

Ordenar los registros de un `DataFrame` por múltiples columnas.

## Solución:

In [33]:
df_array.replace(1600,710,inplace=True)

In [34]:
df_array.sort_values(by=['x','y','z'], ascending=[True,False,True])

x     y     z     u     v
g   504   371   433  1064  1968
a   710   710   980  1347  1537
d   857   342  1218   650  1610
f   916  1144   452   942   414
b  1581   243   114   940  1068
c  1699  1470   196  1023   811
e  1728  1476   129   568   226
e  1881   774   347  1232   495

# Ejercicio 152

Obtener los `n` registros menores de un `DataFrame` con `nsmallest`.

## Solución:

In [36]:
df_array.nsmallest(3,'x') # LAs 3 filas con los menores valores de x


x    y     z     u     v
g  504  371   433  1064  1968
a  710  710   980  1347  1537
d  857  342  1218   650  1610

# Ejercicio 159

Obtener la posición numérica de una columna a partir de su nombre con `get_loc`.

## Solución:

In [41]:
df.columns.get_loc('employ_id')

0

# Ejercicio 160

Contar el número de columnas que tiene un `DataFrame`.

## Solución:

In [46]:
len(df.columns)

11